File: server/apps/api-service/src/main/java/com/sharetimer/apiservice/adapter/in/web/dto/TimerInfoRes.java
package com.sharetimer.apiservice.adapter.in.web.dto;

import java.time.Instant;
import java.util.List;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import io.swagger.v3.oas.annotations.media.Schema;


/**
 * Timer info response
 */
@Schema(name = "Timer Info Response", description = "Timer info response")
@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
public record TimerInfoRes(

    @JsonProperty("updatedAt") @Schema(name = "updatedAt", description = "Timer update time (UTC)",
        example = "2025-07-26T15:00:00Z") @JsonFormat(shape = JsonFormat.Shape.STRING,
            pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", timezone = "UTC") Instant updatedAt,

    @JsonProperty("targetTime") @Schema(name = "targetTime",
        description = "Timer's target time (UTC)",
        example = "2025-07-26T15:00:00Z") @JsonFormat(shape = JsonFormat.Shape.STRING,
            pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", timezone = "UTC") Instant targetTime,

    @JsonProperty("serverTime") @Schema(name = "serverTime", description = "Server time (UTC)",
        example = "2025-07-26T15:00:00Z") @JsonFormat(shape = JsonFormat.Shape.STRING,
            pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", timezone = "UTC") Instant serverTime,

    @JsonProperty("timestamps") @Schema(name = "timestamps",
        description = "List of timer timestamps") List<TimestampInfoRes> timestamps,

    @JsonProperty("isOwner") @Schema(name = "isOwner", description = "Is Owner",
        example = "true") boolean isOwner

) {
}

File: server/apps/api-service/src/main/kotlin/com/sharetimer/apiservice/adapter/input/web/dto/TimerInfoRes.kt
package com.sharetimer.apiservice.adapter.input.web.dto

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty
import com.sharetimer.apiservice.domain.model.Timer
import io.swagger.v3.oas.annotations.media.Schema
import java.time.Instant

/**
 * Timer info response
 */
@Schema(name = "Timer Info Response", description = "Timer info response")
data class TimerInfoRes(
    @field:JsonProperty("updatedAt")
    @field:Schema(
        name = "updatedAt",
        description = "Timer update time (UTC)",
        example = "2025-07-26T15:00:00Z",
    )
    @field:JsonFormat(
        shape = JsonFormat.Shape.STRING,
        pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'",
        timezone = "UTC",
    )
    val updatedAt: Instant?,
    @field:JsonProperty("targetTime")
    @field:Schema(
        name = "targetTime",
        description = "Timer's target time (UTC)",
        example = "2025-07-26T15:00:00Z",
    )
    @field:JsonFormat(
        shape = JsonFormat.Shape.STRING,
        pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'",
        timezone = "UTC",
    )
    val targetTime: Instant?,
    @field:JsonProperty("serverTime")
    @field:Schema(
        name = "serverTime",
        description = "Server time (UTC)",
        example = "2025-07-26T15:00:00Z",
    )
    @field:JsonFormat(
        shape = JsonFormat.Shape.STRING,
        pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'",
        timezone = "UTC",
    )
    val serverTime: Instant?,
    @field:JsonProperty("timestamps")
    @field:Schema(name = "timestamps", description = "List of timer timestamps")
    val timestamps: List<TimestampInfoRes>,
    @field:JsonProperty("isOwner")
    @field:Schema(name = "isOwner", description = "Is Owner", example = "true")
    val isOwner: Boolean,
) {
    companion object {
        fun from(
            serverTime: Instant,
            isOwner: Boolean,
            timer: Timer,
        ): TimerInfoRes =
            TimerInfoRes(
                updatedAt = timer.updatedAt,
                targetTime = timer.targetTime,
                serverTime = serverTime,
                timestamps = TimestampInfoRes.fromList(timer.timestamps),
                isOwner = isOwner,
            )
    }
}
